### PR TITLE
Лабораторна робота №4

### PR DESCRIPTION
**Pull request by MariaShepelieva**
<h1>Дерева рішень</h1>

<h2>Imports</h2>
В рамках цієї лабораторної роботи ми маємо такі імпорти:

```
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, GridSearchCV, KFold, validation_curve
from sklearn.metrics import  accuracy_score, accuracy_score, classification_report
from sklearn.tree import export_graphviz, DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
import pydotplus
```

<h2>Зчитування-інформація-обробка даних</h2>

В результаті цих дій отримали наступні outputs:
image
image
image
image

<h2>Розподіл значень цільової змінної (y)</h2>

image

Розподіл нерівномірний, є явне переважання значення цільової змінної "yes". Змістовно це добрі показники для нашого банку.

<h2>Розбиття  набору даних</h2>

image

Розбиття: 80% на 20%

<h2>DecisionTreeClassifier()</h2>

Оцінка для класифікації "Деревом рішень" показала значення: **0.6756** . Значення дає нам зрозуміти, що для нашаої моделі "Дерево рішень" не є бажано хорошою в прогнозуванні методикою. Тобто точність моделі на тестовій вибірці складає 67.56%.

<h1>Налаштування гіперпараметрів дерева</h1>

<h2>Генератор розбиття</h2>
`kf = KFold( n_splits = 5, shuffle = True, random_state = 42 )`

<h2>Крос-валідація підбору гіперпараметрів</h2>

```
param_grid = {
    'max_depth': [3, 5, 10, 15, 20, None],
    'min_samples_split': [2, 5],
    'min_samples_leaf': [1, 2],
    'max_features': [None, 'sqrt', 'log2']  
}

clf = GridSearchCV(DecisionTreeClassifier(random_state=42), param_grid, cv=kf, scoring='accuracy')
clf.fit(X_train, y_train)
print(f"Найкращі параметри: {clf.best_params_}")
print(f"Точність GridSearchCV : {clf.best_score_:.4f}")
```

задавши інтервали зміни гіперпараметрів вручну ми змогли проаналізувати їх, знайти найкращі параметри для тестової вибірки і  за допомогою GridSearchCV побачити точність вибору гіперпараметрів.

```
Найкращі параметри: {'max_depth': 3, 'max_features': None, 'min_samples_leaf': 1, 'min_samples_split': 2}
Точність GridSearchCV : 0.7739
```

<h2>Валідаційні криві</h2>
Далі за допомогою валідаційних кривих побудувала графіки залежностей метрики від значень гіперпараметрів і в результаті отримала наступні рисунки:

image
image
image
image

Проаналізувавши кожен рисунок ми бачимо, що підтверджується висновок про найкращі параметри, а саме : **max_depth = 3, max_features = None, min_samples_leaf =  1, min_samples_split = 2**
Не варто забувати, що ці значення направляють нас на побудову дерева рішень. Глянемо якого саме:

<h2>Дерево рішень</h2>

image

Важливість ознак: 

image

ппідтверджує мої припущення в попередній роботі, що 'duration' - найважливіша факторна ознака.

<h1>Випадковий ліс</h1> 
<h2>Гіперпараметри та оцінка</h2>
Здійснивши підбір гіперпараметрів моделі:

```
param_ranges = {

    "max_depth": [3, 5, 10, 15, 20, None],
    "min_samples_split": [2, 3, 4, 5], 
    "n_estimators": [10, 20, 30, 65, 80],
}
```

та оцінивши точність класифікації Випадкового лісу, отримали значення: **0.7595** , або ж 75.95% . Тобто можемо зробити висновок, що для нашої моделі метод Випадкового лісу є значно кращим варіантом, але знову ж таки не ідеальним. 

<h2>Валідаційні криві</h2>

Спробуємо цього разу визначити найкращі параметри за допомогою графіків для кожного з гіперпараметру.

image
image
image
image

Проаналізувавши, робимо висновок про настпуні параметри, як "найкращі": **max_depth = 3, max_features = None, min_samples_split = 2, n_estimators = 80** .

Найкорисніші ознаки:

image

Бачимо, що висновки в моїй попередній лабораторній підтвердились даним зображенням. Дійсно, 'duration' має найбільшу важливість оцінки,  а також є найкориснішою ознакою нашого набору даних для target_value.

<h2>Порівняння</h2>

image

Висновки з порівняння такі:
1)  Random Forest показує найкращу загальну точність серед представлених моделей.
2)  KNN краще класифікує об'єкти класу 1, ніж класу 0, показуючи значно вищі метрики Precision, Recall та F1-score для класу 1.
3) Random Forest демонструє найкращі показники F1-score для обох класів, особливо класу 1. Це підтверджує, що Random Forest є найефективнішою моделлю серед представлених.

**Отже:**
Random Forest забезпечує найвищу точність (78.24%) та найкращі метрики Precision, Recall і F1-score для класифікації, особливо для класу 1. KNN також показує гарний результат (74.05%), тоді як Decision Tree поступається іншим моделям за всіма показниками.

Висновок: Треба використовувати Random Forest для заданої задачі класифікації.

### CODE REVIEW

**Comment by MariaShepelieva** [comment]
> # 2 Налаштування гіперпараметрів дерева
> 
> 1.  Створіть генератор розбиття, який перемішує вибірку перед створенням блоків (shuffle=True). Число блоків n_splits дорівнює >5 . Вкажіть параметр random_state для відтворення результатів. Наприклад:
`kf = KFold(n_splits=5, shuffle=True, random_state=42)`

> 2. Здійсніть крос-валідацію моделі для підбору гіперпараметрів: максимальна глибина дерева max_depth;мінімальна кількість об’єктів для розбиття у внутрішній вершині min_samples_split; мінімальна кількість об’єктів у аркуші min_samples_leaf; максимальна кількість ознак, що розглядаються при пошуку кращого розбиття max_features.
> Використовуйте GridSearchCV. Інтервали зміни гіперпараметрів задайте самостійно.

Отримали: 

```
Точність класифікації Дерева рішень (accuracy): 0.8720
Найкращі параметри: {'max_depth': 3, 'max_features': None, 'min_samples_leaf': 1, 'min_samples_split': 2}
Точність GridSearchCV : 0.8997
```

### Висновки

**Оптимальні гіперпараметри:**
- **max_depth=3**: Невелика глибина дерева допомагає уникнути перенавчання.
- **max_features=None**: Використовуються всі доступні ознаки, що забезпечує максимальну інформативність.
- **min_samples_leaf=1**: Кожен лист містить щонайменше один об'єкт.
- **min_samples_split=2**: Для поділу вузла потрібно щонайменше два об'єкти.

**Результати:**
- Точність моделі на тестових даних склала **89.97%**.

**Аналіз:**
- Модель демонструє хорошу узгодженість із даними, без ознак перенавчання.
- Використання всіх ознак дозволяє досягти високої інформативності та точності.

---


> Побудуйте валідаційні криві (залежність метрики від значень гіперпараметрів). Зробіть висновки.

image

image

image

### Висновки

На основі трьох графіків можна зробити такі висновки:

1. **max_depth (глибина дерева):**
   - Чим більша глибина дерева (`max_depth`), тим краще модель адаптується до навчальної вибірки (accuracy для навчання збільшується).
   - Проте на тестовій вибірці спостерігається зниження точності з ростом глибини, що свідчить про **перенавчання**. 
   - Оптимальна глибина знаходиться в діапазоні 5–10, де баланс між навчальною та тестовою вибіркою кращий.

2. **min_samples_split (мінімальна кількість зразків для розбиття вузла):**
   - Збільшення значення `min_samples_split` зменшує точність навчальної вибірки, оскільки модель стає менш складною.
   - Точність тестової вибірки злегка зростає при підвищенні `min_samples_split`, що свідчить про зменшення переобучення.
   - Значення `min_samples_split` близько 4–5 виглядає оптимальним.

3. **min_samples_leaf (мінімальна кількість зразків у листі):**
   - При збільшенні `min_samples_leaf` точність на навчальній вибірці зменшується, але на тестовій вибірці поступово зростає.
   - Це також вказує на зменшення переобучення, і значення `min_samples_leaf` в межах 4–6 є оптимальним компромісом.

**Загальний висновок:** 
- Для досягнення оптимальної точності потрібно налаштувати параметри:
  - **max_depth** =3,
  - **min_samples_split** =20,
  - **min_samples_leaf** =10.
- Збільшення складності дерева (глибина) чи зменшення мінімальної кількості зразків у вузлах може погіршити точність через перенавчання.

--- 

> 4. Побудуйте графічно отримане дерево. Оцініть важливість ознак за допомогою атрибуту feature_importances_.
> 
> 5. Сформулюйте висновки.

image
image

### Висновки на основі аналізу дерева та важливості ознак

1. **Аналіз структури дерева**
   - Модель використала лише кілька найбільш важливих ознак для поділу, що свідчить про їх значний вплив на результат.
   - Основні вузли дерева включають:
     - `nr.employed`: ключова змінна, яка була використана для першого розбиття.
     - `cons.conf.idx`, `pdays`, `euribor3m`: ці ознаки також часто зустрічаються у головних розгалуженнях дерева.
   - Глибина дерева залишається обмеженою, що допомагає зменшити ризик перенавчання.

2. **Аналіз важливості ознак**
   - Найважливіша ознака: 
     - `nr.employed`, яка має найвищий вплив.
   - Важливі, але другорядні ознаки:
     - `pdays` і `cons.conf.idx`.
   - Менш значущі, але помітні ознаки:
     - `month_oct`, `euribor3m`, `contact_telephone`.
   - Інші ознаки мають незначний вплив, тому їх можна буде виключити без втрати точності моделі.

---
---

> # 3  Випадковий ліс
> ---
> 1. Побудуйте модель випадкового лісу (Random Forest) для вашої задачі з гіперпараметрами за замовчуванням та оцініть якість моделі на відкладеній вибірці.

Результат: 
```

Random Forest:        0.8993

=== Звіт для Random Forest ===
              precision    recall  f1-score   support

           0       0.91      0.98      0.95     10968
           1       0.65      0.23      0.34      1389

   macro avg       0.78      0.61      0.64     12357
weighted avg       0.88      0.90      0.88     12357
```

Аналіз моделі випадкового лісу  

1. **Загальна ефективність моделі**  
Модель має точність 89.28%. Вона чудово справляється з класифікацією основного класу ("No"), але її продуктивність для рідкісного класу ("Yes") залишається низькою.  

2. **Результати для класу "No"**  
- **Точність (Precision)**: 0.92 — модель рідко помиляється при прогнозуванні цього класу.  
- **Повнота (Recall)**: 0.97 — майже всі об’єкти класу "No" правильно визначаються.  
- **F1-score**: 0.94 — хороший баланс між точністю та повнотою.  

3. **Результати для класу "Yes"**  
- **Точність (Precision)**: 0.54 — значна кількість помилкових передбачень для цього класу.  
- **Повнота (Recall)**: 0.29 — модель пропускає більшість об’єктів класу "Yes".  
- **F1-score**: 0.38 — низький загальний показник через слабку точність і повноту.  

4. **Середні значення метрик**  
- **Середнє по класах (Macro average)**:  
  - **F1-score**: 0.66 — усереднений показник для обох класів, не враховуючи дисбаланс.  
- **Зважене середнє (Weighted average)**:  
  - **F1-score**: 0.88 — середній показник, скоригований на частку кожного класу в даних.  

5. **Ключові висновки**  
Модель добре працює з основним класом ("No"), однак її продуктивність для класу "Yes" залишає бажати кращого. Дисбаланс у розподілі класів значно впливає на результати. Для задач, де критично важливо коректно класифікувати клас "Yes", модель потребує додаткової оптимізації.

---

> 2. Здійсніть підбір гіперпараметрів моделі: кількість дерев n_estimators; максимальна глибина дерева max_depth; мінімальна кількість об’єктів для розбиття у внутрішній вершині min_samples_split; мінімальна кількість об’єктів у аркуші min_samples_leaf; максимальна кількість ознак, що розглядаються при пошуку кращого розбиття max_features.

```
param_ranges = {

    "max_depth": [3, 5, 10, 15, 20, None],
    "min_samples_split": [2, 3, 4, 5], 
    "n_estimators": [10, 20, 30, 65, 80],
}
```

Отримали результатом значення: 0.7595 , або ж 75.95% . Тобто можемо зробити з цього висновок, що для нашої моделі метод Випадкового лісу є значно кращим варіантом, але знову ж таки не ідеальним.

---

> 3. Побудуйте валідаційні криві для кожного з гіперпараметрів

image

image

image

image


Висновки:

1. **Вплив параметра `max_depth`**  
  - **Навчальна вибірка**: Точність зростає зі збільшенням max_depth, оскільки більш глибокі дерева краще пристосовуються до даних.  
  - **Тестова вибірка**: Точність досягає піку при max_depth у межах 5–10, а потім знижується через перенавчання.  

**Висновок**: Оптимальне значення max_depth — 5–10 для забезпечення балансу між адаптацією до даних і узагальненням.  

2. **Вплив параметра `min_samples_split`**  
  - **Навчальна вибірка**: Зі збільшенням min_samples_split точність знижується, оскільки дерево стає менш складним.  
  - **Тестова вибірка**: Точність стабільна (~0.9), що свідчить про меншу залежність тестової вибірки від цього параметра.  

**Висновок**: Значення min_samples_split > 3–4 допомагає уникнути перенавчання.  


3. **Вплив параметра `n_estimators`**  
  - **Навчальна вибірка**: Точність зростає, але після 50 дерев приріст стає незначним.  
  - **Тестова вибірка**: Точність стабільна, але нижча за навчальну, що може свідчити про недостатню узагальненість.  

**Висновоак**: Використання n_estimators ≥ 50 є оптимальним.  

4. **Вплив параметра `max_features`**  
  - **Навчальна вибірка**: Найвища точність при max_features=None, проте це може спричинити перенавчання. Зменшення max_features (до sqrt або log2) знижує точність, але покращує узагальнення.  
  - **Тестова вибірка**: Точність стабільна незалежно від max_features, але розрив між навчальною і тестовою вибірками зменшується при зменшенні цього параметра.  

**Висновок**: Використання max_features=sqrt або log2 допомагає збалансувати точність і час навчання.  

**Загальний висновок**  
Для оптимізації моделі RandomForestClassifier:  
  - **max_depth**: 5–10  
  - **min_samples_split**: 3–4  
  - **n_estimators**: ≥ 50  
  - **max_features**: sqrt або log2  

---

> 4. Оцініть важливість ознак цієї моделі. Візуалізуйте топ-10 найкорисніших ознак за допомогою стовпчастої діаграми.

image

### Висновки щодо впливу ознак у моделі Random Forest:  
- Ознаки, які характеризують клієнта (наприклад, age, housing, loan), мають суттєвий вплив на результати моделі.  
- Економічні показники (euribor3m, cons.conf.idx, cons.price.idx) відіграють важливу роль у передбаченнях.  
- Фактори, пов’язані з результатами маркетингових кампаній (campaign, poutcome_success), також впливають на модель, але їхній вплив є дещо менш значущим.
